File: packages/server/src/repowise/server/schemas/blast_radius.py
"""Blast-radius request/response models."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from repowise.server.schemas.risk_semantics import (
    RiskCompatibilityField,
    RiskScalarSemantics,
)


class BlastRadiusRequest(BaseModel):
    changed_files: list[str]
    max_depth: int = Field(default=3, ge=1, le=10)


class DirectRiskEntry(BaseModel):
    path: str
    structural_score: float
    #: Deprecated exact alias of ``structural_score`` for older clients.
    risk_score: float
    temporal_hotspot: float
    centrality: float


class TransitiveEntry(BaseModel):
    path: str
    depth: int


class CochangeWarning(BaseModel):
    changed: str
    missing_partner: str
    score: float


class ReviewerEntry(BaseModel):
    email: str
    files: int
    ownership_pct: float


class TestImpactEvidence(BaseModel):
    basis: Literal["measured", "inferred"]
    source_file: str
    via: Literal["coverage-map", "call-graph", "import-graph"]
    source_format: str | None = None


class TestRecommendation(BaseModel):
    test_id: str
    test_file: str | None = None
    repository_id: str
    repository: str
    basis: Literal["measured", "inferred"]
    bases: list[Literal["measured", "inferred"]]
    source_files: list[str]
    evidence: list[TestImpactEvidence]


class TestImpactFile(BaseModel):
    source_file: str
    status: Literal["measured", "inferred", "unknown"]
    measured_tests: list[str]
    measured_tests_total: int
    inferred_tests: list[str]
    inferred_tests_total: int


class TestImpactFreshness(BaseModel):
    status: Literal["current", "stale", "unknown"]
    reason: str | None = None
    ingested_commit: str | None = None
    indexed_commit: str | None = None


class TestImpactCoverage(BaseModel):
    status: Literal["available", "partial", "unavailable", "degraded"]
    reason: str | None = None
    map_present: bool
    pair_count: int
    test_count: int
    source_file_count: int
    changed_files_total: int
    changed_files_with_measured_tests: int
    changed_files_without_measured_tests: int
    ingested_at: str | None = None
    source_format: str | None = None
    freshness: TestImpactFreshness


class TestImpactInference(BaseModel):
    status: Literal["available", "degraded"]
    reason: str | None = None
    changed_files_total: int
    changed_files_with_candidates: int
    candidates_before_dedup: int


class TestImpactAnalysis(BaseModel):
    status: Literal["available", "partial", "degraded"]
    stale: bool
    partial: bool
    degraded: bool
    basis_categories: list[Literal["measured", "inferred"]]


class TestImpactResponse(BaseModel):
    recommendations: list[TestRecommendation]
    recommendations_total: int
    recommendations_emitted: int
    recommendations_truncated: bool
    recommendations_omitted: int
    recommendations_by_primary_basis: dict[Literal["measured", "inferred"], int]
    files: list[TestImpactFile]
    files_total: int
    files_without_measured_tests: list[str]
    unknown_files: list[str]
    coverage: TestImpactCoverage
    inference: TestImpactInference
    analysis: TestImpactAnalysis


class BlastRadiusResponse(BaseModel):
    direct_risks: list[DirectRiskEntry]
    transitive_affected: list[TransitiveEntry]
    cochange_warnings: list[CochangeWarning]
    recommended_reviewers: list[ReviewerEntry]
    test_gaps: list[str]
    test_impact: TestImpactResponse
    structural_impact_score: float
    structural_impact_band: Literal["localized", "moderate", "broad"]
    structural_impact_scale: RiskScalarSemantics
    #: Deprecated exact alias; never repurposed to a new unit.
    overall_risk_score: float
    overall_risk_score_compatibility: RiskCompatibilityField

File: packages/types/src/blast-radius.ts
/**
 * Blast-radius types — mirror the OSS engine's `BlastRadius*` schemas
 * (`packages/server/src/repowise/server/schemas.py`) and the hosted
 * backend's `app/models/schemas.py` `BlastRadius*` Pydantic models.
 *
 * Both backends produce the same shape; this is the canonical TS contract.
 */

import type {
  RiskCompatibilityField,
  RiskScalarSemantics,
} from "./risk-semantics.js";

export interface DirectRiskEntry {
  path: string;
  /** Raw unbounded structural heuristic; compare only within this change set. */
  structural_score: number;
  /** @deprecated Exact alias of structural_score for older clients. */
  risk_score: number;
  /** 0-1 normalised. */
  temporal_hotspot: number;
  /** Raw graph centrality (pagerank); typically well below 1. */
  centrality: number;
}

export interface TransitiveEntry {
  path: string;
  depth: number;
}

export interface CochangeWarning {
  changed: string;
  missing_partner: string;
  /** Co-change frequency count from git history. */
  score: number;
}

export interface ReviewerEntry {
  email: string;
  files: number;
  /** 0–1 fraction. UI multiplies by 100 for display. */
  ownership_pct: number;
}

export type TestRecommendationBasis = "measured" | "inferred";

export interface TestImpactEvidence {
  basis: TestRecommendationBasis;
  source_file: string;
  via: "coverage-map" | "call-graph" | "import-graph";
  source_format: string | null;
}

export interface TestRecommendation {
  test_id: string;
  test_file: string | null;
  repository_id: string;
  repository: string;
  /** Strongest evidence on this de-duplicated recommendation. */
  basis: TestRecommendationBasis;
  /** Every retained evidence category; measured and inferred may both occur. */
  bases: TestRecommendationBasis[];
  source_files: string[];
  evidence: TestImpactEvidence[];
}

export interface TestImpactResponse {
  recommendations: TestRecommendation[];
  recommendations_total: number;
  recommendations_emitted: number;
  recommendations_truncated: boolean;
  recommendations_omitted: number;
  /** Exact, mutually exclusive totals over the full recommendation population. */
  recommendations_by_primary_basis: Record<TestRecommendationBasis, number>;
  files: Array<{
    source_file: string;
    status: "measured" | "inferred" | "unknown";
    measured_tests: string[];
    measured_tests_total: number;
    inferred_tests: string[];
    inferred_tests_total: number;
  }>;
  files_total: number;
  files_without_measured_tests: string[];
  unknown_files: string[];
  coverage: {
    status: "available" | "partial" | "unavailable" | "degraded";
    reason: string | null;
    map_present: boolean;
    pair_count: number;
    test_count: number;
    source_file_count: number;
    changed_files_total: number;
    changed_files_with_measured_tests: number;
    changed_files_without_measured_tests: number;
    ingested_at: string | null;
    source_format: string | null;
    freshness: {
      status: "current" | "stale" | "unknown";
      reason: string | null;
      ingested_commit: string | null;
      indexed_commit: string | null;
    };
  };
  inference: {
    status: "available" | "degraded";
    reason: string | null;
    changed_files_total: number;
    changed_files_with_candidates: number;
    candidates_before_dedup: number;
  };
  analysis: {
    status: "available" | "partial" | "degraded";
    stale: boolean;
    partial: boolean;
    degraded: boolean;
    basis_categories: TestRecommendationBasis[];
  };
}

export interface BlastRadiusResponse {
  direct_risks: DirectRiskEntry[];
  transitive_affected: TransitiveEntry[];
  cochange_warnings: CochangeWarning[];
  recommended_reviewers: ReviewerEntry[];
  test_gaps: string[];
  /** Canonical typed test-impact population. Optional for older servers. */
  test_impact?: TestImpactResponse;
  /** Deterministic, uncalibrated structural-impact heuristic, 0–10. */
  structural_impact_score: number;
  structural_impact_band: "localized" | "moderate" | "broad";
  structural_impact_scale: RiskScalarSemantics;
  /** @deprecated Exact alias of structural_impact_score; not a probability. */
  overall_risk_score: number;
  overall_risk_score_compatibility: RiskCompatibilityField;
}

export interface BlastRadiusRequest {
  changed_files: string[];
  max_depth?: number;
}
